yjit/src/compilation_log.rs

```diff
@@ -0,0 +1,180 @@
+use crate::core::BlockId;
+use crate::cruby::*;
+use crate::options::*;
+use crate::yjit::yjit_enabled_p;
+
+use std::fmt::{Display, Formatter};
+use std::os::raw::c_long;
+use crate::utils::iseq_get_location;
+
+type Timestamp = f64;
+
+#[derive(Clone, Debug)]
+pub struct CompilationLogEntry {
+    /// The time when the block was compiled.
+    pub timestamp: Timestamp,
+
+    /// The compilation log message.
+    pub message: String,
+}
+
+impl Display for CompilationLogEntry {
+    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
+        write!(f, "{:15.6}: {}", self.timestamp, self.message)
+    }
+}
+
+pub type CompilationLog = CircularBuffer<CompilationLogEntry, 1024>;
+static mut COMPILATION_LOG : Option<CompilationLog> = None;
+
+impl CompilationLog {
+    pub fn init() {
+        unsafe {
+            COMPILATION_LOG = Some(CompilationLog::new());
+        }
+    }
+
+    pub fn get_instance() -> &'static mut CompilationLog {
+        unsafe {
+            COMPILATION_LOG.as_mut().unwrap()
+        }
+    }
+
+    pub fn has_instance() -> bool {
+        unsafe {
+            COMPILATION_LOG.as_mut().is_some()
+        }
+    }
+
+    pub fn add_block_with_chain_depth(block_id: BlockId, chain_depth: u8) {
+        if !Self::has_instance() {
+            return;
+        }
+
+        let print_compilation_log = get_option!(compilation_log);
+        let timestamp = std::time::SystemTime::now().duration_since(std::time::UNIX_EPOCH).unwrap().as_secs_f64();
+
+        let location = iseq_get_location(block_id.iseq, block_id.idx);
+        let index = block_id.idx;
+        let message = if chain_depth > 0 {
+            format!("{} (index: {}, chain_depth: {})", location, index, chain_depth)
+        } else {
+            format!("{} (index: {})", location, index)
+        };
+
+        let entry = CompilationLogEntry {
+            timestamp,
+            message
+        };
+
+        if let Some(output) = print_compilation_log {
+            match output {
+                CompilationLogOutput::Stderr => {
+                    eprintln!("{}", entry);
+                }
+
+                CompilationLogOutput::File(fd) => {
+                    use std::os::unix::io::{FromRawFd, IntoRawFd};
+                    use std::io::Write;
+
+                    // Write with the fd opened during boot
+                    let mut file = unsafe { std::fs::File::from_raw_fd(fd) };
+                    writeln!(file, "{}", entry).unwrap();
+                    file.flush().unwrap();
+                    file.into_raw_fd(); // keep the fd open
+                }
+
+                CompilationLogOutput::MemoryOnly => () // Don't print or write anything
+            }
+        }
+
+        Self::get_instance().push(entry);
+    }
+}
+
+pub struct CircularBuffer<T, const N: usize> {
+    buffer: Vec<Option<T>>,
```

Comment:
```suggestion
    buffer: [T; N],
```

An array here should work just fine.

Reply:
Thanks. I think the iterator was getting in the way because `String` isn't `Copy`. I like this better.

Why do you suggest getting rid of the `Option`? I've been using `Option<T>#take` to clear the value once read. I think it makes `pop()` look nicer.

Reply:
Oh sorry, I meant `[Option<T>; N]`. That should make it easy to initialize too right?

Reply:
nvm, this is not a big issue, and in the past we've ran into footguns with accidentally creating big arrays on the stack. Let's just leave it as is. https://github.com/rust-lang/rust/issues/53827